packages/plugin-core/src/test/suite-integ/MoveHeader.test.ts

```diff
@@ -0,0 +1,151 @@
+import { DVault, NoteProps } from "@dendronhq/common-all";
+import {
+  NoteTestUtilsV4,
+  PreSetupHookFunction,
+} from "@dendronhq/common-test-utils";
+import { VSCodeUtils } from "../../utils";
+import { runLegacyMultiWorkspaceTest, setupBeforeAfter } from "../testUtilsV3";
+import { describe, beforeEach } from "mocha";
+import vscode from "vscode";
+import { expect } from "../testUtilsv2";
+import { MoveHeaderCommand } from "../../commands/MoveHeader";
+import _ from "lodash";
+
+suite("MoveHeader", function () {
+  const ctx = setupBeforeAfter(this, {});
+
+  describe("GIVEN a note with a simple header", () => {
+    let originNote: NoteProps;
+    let destNote: NoteProps;
+    let preSetupHook: PreSetupHookFunction;
+
+    beforeEach(() => {
+      preSetupHook = async ({
+        wsRoot,
+        vaults,
+      }: {
+        wsRoot: string;
+        vaults: DVault[];
+      }) => {
+        originNote = await NoteTestUtilsV4.createNote({
+          fname: "origin",
+          wsRoot,
+          vault: vaults[0],
+          body: "## Foo header\n\n",
+        });
+        destNote = await NoteTestUtilsV4.createNote({
+          fname: "dest",
+          wsRoot,
+          vault: vaults[0],
+          body: "# Some header",
+        });
+        await NoteTestUtilsV4.createNote({
+          fname: "ref-note",
+          wsRoot,
+          vault: vaults[0],
+          body: "[[Foo|origin#foo-header]]",
+        });
+      };
+    });
+
+    describe("WHEN header is selected", () => {
+      let onInitFunc: Function;
+      beforeEach(() => {
+        onInitFunc = (nextFunc: Function) => {
+          return async () => {
+            const editor = await VSCodeUtils.openNote(originNote);
+            editor.selection = new vscode.Selection(7, 0, 7, 0);
+            nextFunc();
+          };
+        };
+      });
+
+      test("THEN selected header is moved from origin to dest", (done) => {
+        runLegacyMultiWorkspaceTest({
+          ctx,
+          preSetupHook,
+          onInit: onInitFunc(async () => {
+            const cmd = new MoveHeaderCommand();
+            const out = await cmd.run({ dest: destNote });
+            expect(out!.origin.body.includes("## Foo header")).toBeFalsy();
+            expect(out!.dest!.body.includes("## Foo header")).toBeTruthy();
+            done();
+          }),
+        });
+      });
+
+      test("THEN reference to moved header is updated", (done) => {
+        runLegacyMultiWorkspaceTest({
```

Comment:
If I am not mistaken this `runLegacyMultiWorkspaceTest` is doing brand new workspace setup each time that it runs, some concern whether this will make our tests slow and whether we want to move such setup higher up and into beforeAll?

Reply:
The problem here is that these testing utilities aren't modular. I can probably make it _slightly_ more readable by currying `runLegacyMultiWorkspaceTest` on each step of the Given-When-Then clauses, but at the end it will be invoked once all information needed to test is there, which is what I'm currently doing here already.

This is a pretty substantial friction at this point because we pretty much have to undo what `runLegacyMultiWorkspaceTest` is bundling together into individual lifecycles, and then spread those on top of the new BDD-light style structure of testing.

Not saying it couldn't be done or shouldn't be done. I'm just a bit concerned if we should focus on it _now_. Was curious if you had any thoughts on it.

Reply:
@SeriousBug is working on native workspaces. as part of that, he'll be writing a new test harness to setup the native workspaces. we should make sure the new test harness has the desired properties that make sense. created an rfc/discussion [here](https://github.com/dendronhq/dendron/discussions/1394)